Add XCTS Schwarzschild solution in more coords

nilsvu opened the pull request:
## Proposed changes

Add Painlevé-Gullstrand and isotropic Kerr-Schild coords.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add Xcts::Solutions namespace to docs

nilsvu committed:
Rename a test file

nilsvu committed:
Remove suffixes in test file

nilsvu committed:
Add XCTS Schwarzschild solution in more coords

Add Painlevé-Gullstrand and isotropic Kerr-Schild coords.

nilsvu committed:
Disambiguate a Tags namespace

wthrowe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Xcts/SchwarzschildIsotropic.py:
```diff
@@ -2,55 +2,50 @@
 # See LICENSE.txt for details.
```
Commit message "prefixes" -> "suffixes"

wthrowe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Xcts/SchwarzschildKerrSchildIsotropic.py:
```diff
@@ -0,0 +1,209 @@
+
+
+# Isotropic Kerr-Schild coordinates
+# See e.g. Sec. 7.4.1 in Harald's thesis
```
Harald who? :)

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/Xcts/Schwarzschild.cpp:
```diff
@@ -44,17 +52,69 @@ Options::create_from_yaml<Xcts::Solutions::SchwarzschildCoordinates>::create<
+  return isotropic_radius / areal_radius * one_over_lapse;
+}
+
+double kerr_schild_areal_radius_from_isotropic(const double isotropic_radius,
```
Codecov reports this function as untested.  If that is correct, test it.

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/Xcts/Schwarzschild.cpp:
```diff
@@ -143,72 +252,240 @@ template <typename DataType>
+      const auto one_over_lapse = sqrt(1. + 2. * mass / r);
+      const DataType isotropic_prefactor = -conformal_factor * mass / r /
+                                           (1. + one_over_lapse) /
+                                           one_over_lapse / square(rbar);
```
Multiply in the denominator instead of repeatedly dividing.

nilsvu replied:
Is that generally better?

wthrowe replied:
Division is sloooow.  Several times as expensive as multiplication, typically.

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/Xcts/Schwarzschild.cpp:
```diff
@@ -226,17 +503,86 @@ void SchwarzschildVariables<DataType>::operator()(
+          get(cache->get_var(*this, detail::Tags::ArealRadius<DataType>{}));
+      const auto& lapse =
+          get(cache->get_var(*this, gr::Tags::Lapse<DataType>{}));
+      const DataType isotropic_prefactor = 2. * mass * lapse / square(r);
```
The documentation says lapse^2.  Which is correct?

nilsvu replied:
No square. It's Eq. (19) in the docs, at the bottom of the KerrSchildIsotropic docs, which is correct I think.

wthrowe replied:
You are correct.  I was reading the wrong equation.

nilsvu commented:
Thanks for your review @wthrowe !

wthrowe commented:
Squash.

nilsvu commented:
@macedo22 would you mind doing the second review for this?

macedo22 commented on src/PointwiseFunctions/AnalyticSolutions/Xcts/Schwarzschild.hpp:
```diff
@@ -68,6 +68,90 @@ enum class SchwarzschildCoordinates {
+   * which has the solution
+   *
+   * \begin{equation}
+   *   \bar{r} = \frac{r}{4} \left(1 + \sqrt{\frac{2M}{r}}\right)^2
```
I think `\sqrt{\frac{2M}{r}}` -> `\sqrt{1 + \frac{2M}{r}}` according to eq 7.34 [here](https://arxiv.org/pdf/gr-qc/0510016.pdf).

nilsvu replied:
yep, thanks for catching this 👍

macedo22 commented on src/PointwiseFunctions/AnalyticSolutions/Xcts/Schwarzschild.cpp:
```diff
@@ -226,17 +521,90 @@ void SchwarzschildVariables<DataType>::operator()(
 template <typename DataType>
 void SchwarzschildVariables<DataType>::operator()(
     const gsl::not_null<tnsr::I<DataType, 3>*> shift_excess,
```
(Optional) From reading the implementation, this seems to just be the shift. I see that there is a `shift_background` and a `longitudinal_shift_background_minus_dt_conformal_metric` that are just initialized to 0. If the `_background` and `_excess` terminology is something GR people readily recognize and understand why we have both, please ignore me, but it wasn't clear to me why we need the two values and what purpose the `shift_background` is serving. If you think it's not overexplaining things or not readily obvious to GR people, it might be helpful to add a comment for the `shift_excess` and `shift_background` functions saying what they are / why both are needed / that `shift_excess` is just the shift (if that's true).

nilsvu replied:
The shift decomposition is an implementation detail of the XCTS system. Added a comment and ref to https://spectre-code.org/structXcts_1_1Tags_1_1ShiftExcess.html.

macedo22 replied:
Ah, I see, then I think it's probably on me that I should've read more documentation first!

macedo22 commented on src/PointwiseFunctions/AnalyticSolutions/Xcts/Schwarzschild.cpp:
```diff
@@ -226,17 +521,90 @@ void SchwarzschildVariables<DataType>::operator()(
-    Tags::ShiftStrain<DataType, 3, Frame::Inertial> /*meta*/) const {
-  std::fill(shift_strain->begin(), shift_strain->end(), 0.);
+    const gsl::not_null<Cache*> cache,
+    Xcts::Tags::ShiftStrain<DataType, 3, Frame::Inertial> /*meta*/) const {
```
(Optional) I just wanted to mention that I wasn't able to verify this implementation because I don't see a tensor equation for this referenced in the documentation. If you think this is something that is not obvious and worth documenting, then I would add it somewhere (e.g. add the equation to the class doc or reference another part of spectre or a textbook eq).

macedo22 commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Xcts/SchwarzschildKerrSchildIsotropic.py:
```diff
@@ -0,0 +1,209 @@
+    lapse = lapse_areal(r, mass)
+    K = extrinsic_curvature_trace_areal(r, mass)
+    return (K / r * (3. * mass * lapse**2 / r - 3. * mass /
+                     (r + 3. * mass) - 2.))
```
I'm not sure if I'm missing something, but according to the cpp implementation, shouldn't this be divided by `square(rbar)`?

macedo22 commented on src/PointwiseFunctions/AnalyticSolutions/Xcts/Schwarzschild.cpp:
```diff
@@ -4,8 +4,10 @@
 #include "PointwiseFunctions/AnalyticSolutions/Xcts/Schwarzschild.hpp"
 
 #include <algorithm>
```
I think `cstddef` is missing for `size_t`

macedo22 commented on src/PointwiseFunctions/AnalyticSolutions/Xcts/Schwarzschild.cpp:
```diff
@@ -22,6 +25,7 @@
 #include "Utilities/ConstantExpressions.hpp"
 #include "Utilities/ErrorHandling/Error.hpp"
 #include "Utilities/Gsl.hpp"
+#include "Utilities/MakeWithValue.hpp"
```
I think `Utilities/TMPL.hpp` is missing for `tmpl::size_t`

macedo22 commented on src/PointwiseFunctions/AnalyticSolutions/Xcts/Schwarzschild.cpp:
```diff
@@ -4,8 +4,10 @@
+#include <limits>
 #include <ostream>
 #include <pup.h>
+#include <string>
```
I think `tuple` is missing for `std::make_tuple`

macedo22 reviewed (changes requested):
LGTM, just a couple nitpicks and suggestions

macedo22 commented:
Feel free to directly squash any changes you decide to make

nilsvu commented:
Thanks @macedo22, I pushed a fixup to give you the chance to verify the changes

macedo22 commented:
LGTM, feel free to squash

nilsvu commented:
Done, thanks for your thorough review @macedo22 !

@wthrowe I think this is ready to go once CI passes

wthrowe commented:
```
The following tests FAILED:
2997
	1271 - Unit.PointwiseFunctions.AnalyticSolutions.Xcts.Schwarzschild (Timeout)
```

Speed the test up or increase the timeout.

nilsvu committed:
Combine some tests to speed them up

The combined call into the analytic solution is faster because it reuses
intermediate computations.

To be safe, also increase the timeout because this test has timed out
on CI.

nilsvu commented:
Added another commit to address this

wthrowe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Xcts/Test_Schwarzschild.cpp:
```diff
@@ -181,6 +148,7 @@ void test_solution(const double mass,
 
 }  // namespace
 
+// [[Timeout, 5]]
```
That's the default.

nilsvu replied:
Set to 8 now

nilsvu commented:
Looks like the test that failed the one time takes around 4 seconds, so I think the timeout of 8 seconds that I set now seems fine. Please take another look when you get the chance @wthrowe.
